### PR TITLE
Formal model for SBFT - initial version

### PR DESCRIPTION
**Pull request by HristoStaykov**
In this version of the model we have the Slow
Commit path, for which we have proven that honest
replicas agree in their Commit messages to the
assignment of Client Operations to Sequence ID-s.
We have the mechanism of the View Change process
in the system modeled according to the definition in
BFT (excluding the Fast Commit Path) because it is
not yet modeled and we are missing the specifics
related to Checkpointing, since they are yet to be
modeled as well.

### CODE REVIEW

File: docs/sbft-formal-model/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+Given the distributed nature of VMware Blockchain, unit tests offer limited means of establishing the correctness of the SMR protocol. The main tool we use for proving correctness is an integration testing framework code-named [Apollo](https://github.com/vmware/concord-bft/tree/master/tests/apollo) which creates mini-blockchain deployments and exercises various runtime scenarios (including straight case, replica failures, network faults, as well as certain byzantine cases).
+
+Apollo does give us good confidence as to the correctness and fault tolerance of the system that we sell to our customers. But even if we bring test coverage to 100%, we can never be sure that we have discovered all possible bugs and failure scenarios, especially in a complex distributed system, such as VMware Blockchain.
+
```
Comment (teoparvanov) [inline docs/sbft-formal-model/README.md:7]: let's skip the "selling" part here in the open source project :)

File: docs/sbft-formal-model/client.i.dfy [hunk 2]
```diff
@@ -0,0 +1,85 @@
+//#title Host protocol
+//#desc Define the host state machine here: message type, state machine for executing one
+//#desc host's part of the protocol.
+
+// See exercise01.dfy for an English design of the protocol.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/client.i.dfy:5]: Let's replace those comments by the Concord-BFT copyright header (2022).

File: docs/sbft-formal-model/application.s.dfy [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+              .(unprocessedRequests := v.unprocessedRequests - {clientRequest})
+  }
+
+  predicate DeliverReply(v:Variables, v':Variables, clientReply:ClientReply)
+  {
+    && clientReply in v.undeliveredReplies
+    && v' == v.(undeliveredReplies := v.undeliveredReplies - {clientReply})
+  }
+
+  // JayNF
```
Comment (teoparvanov) [inline docs/sbft-formal-model/application.s.dfy:50]: Let's be more explicit here: "Jay Normal Form (JNF)". You could also provide a link to an explanation of what it means.

File: docs/sbft-formal-model/client.i.dfy [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  
+  predicate Init(c:Constants, v:Variables) {
+    && v.WF(c)
+    && v.lastRequestTimestamp == 0
+    && v.lastReplyTimestamp == 0
+  }
+
+  // JayNF
+  datatype Step =
+    | SendClientOperationStep()
```
Comment (teoparvanov) [inline docs/sbft-formal-model/client.i.dfy:75]: same as above: "Jay Normal Form (JNF)". You could also provide a link to an explanation of what it means.

File: docs/sbft-formal-model/distributed_system.s.dfy [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    && v.WF(c)
+    && (forall id | && c.clusterConfig.IsReplica(id) 
+                    :: Replica.Init(c.hosts[id].replicaConstants, v.hosts[id].replicaVariables))
+    && (forall id | && c.clusterConfig.IsClient(id)
+                    :: Client.Init(c.hosts[id].clientConstants, v.hosts[id].clientVariables))
+    && Network.Init(c.network, v.network)
+  }
+
+  // JayNF
+  datatype Step = Step(id:HostId, msgOps: Network.MessageOps<Message>)
```
Comment (teoparvanov) [inline docs/sbft-formal-model/distributed_system.s.dfy:70]: same as above: "Jay Normal Form (JNF)". You could also provide a link to an explanation of what it means.

File: docs/sbft-formal-model/cluster_config.s.dfy [hunk 6]
```diff
@@ -0,0 +1,61 @@
+include "network.s.dfy"
```
Comment (teoparvanov) [inline docs/sbft-formal-model/cluster_config.s.dfy:1]: Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/distributed_system.s.dfy [hunk 7]
```diff
@@ -0,0 +1,97 @@
+//#title DistributedSystem
+//#desc DO NOT EDIT THIS FILE! It's a trusted file that specifies how hosts interact
+//#desc with one another over the network.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/distributed_system.s.dfy:3]: Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/messages.dfy [hunk 8]
```diff
@@ -0,0 +1,72 @@
+include "network.s.dfy"
```
Comment (teoparvanov) [inline docs/sbft-formal-model/messages.dfy:1]: Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/network.s.dfy [hunk 9]
```diff
@@ -0,0 +1,87 @@
+//#title Network
+//#desc DO NOT EDIT THIS FILE! It's a trusted file that specifies how the network
+//#desc delivers packets, allowing reorder and duplicate delivery.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/network.s.dfy:3]: Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/proof.dfy [hunk 10]
```diff
@@ -0,0 +1,703 @@
+include "library/Library.dfy"
+include "distributed_system.s.dfy"
```
Comment (teoparvanov) [inline docs/sbft-formal-model/proof.dfy:2]: Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/replica.i.dfy [hunk 11]
```diff
@@ -0,0 +1,524 @@
+//#title Host protocol
+//#desc Define the host state machine here: message type, state machine for executing one
+//#desc host's part of the protocol.
+
+// See exercise01.dfy for an English design of the protocol.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/replica.i.dfy:5]: replace by Concord-BFT copyright header (2022)

File: docs/sbft-formal-model/README.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
+# Project overview
+[SBFT](https://arxiv.org/pdf/1804.01626.pdf) is a byzantine fault-tolerant protocol (BFT) for state machine replication (SMR) designed and implemented in-house at VMware. SBFT is implemented in the [concord-bft](https://github.com/vmware/concord-bft) open source project. It is the foundation on which the [VMware Blockchain](https://docs.vmware.com/en/VMware-Blockchain/1.3/getting_started/GUID-6BD4CD5F-6AC6-4B05-BCB3-A76626BB2777.html) product has been built. A layered architecture enables integration with various smart contract execution engines, the currently supported ones being DAML and Ethereum’s EVM. We can think of VMware Blockchain as a new type of virtualized infrastructure that provides something beyond compute, a human value, which is trust.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/README.md:2]: Let's skip "designed and implemented in-house at VMware"

File: docs/sbft-formal-model/README.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+Given the distributed nature of VMware Blockchain, unit tests offer limited means of establishing the correctness of the SMR protocol. The main tool we use for proving correctness is an integration testing framework code-named [Apollo](https://github.com/vmware/concord-bft/tree/master/tests/apollo) which creates mini-blockchain deployments and exercises various runtime scenarios (including straight case, replica failures, network faults, as well as certain byzantine cases).
+
+Apollo does give us good confidence as to the correctness and fault tolerance of the system that we sell to our customers. But even if we bring test coverage to 100%, we can never be sure that we have discovered all possible bugs and failure scenarios, especially in a complex distributed system, such as VMware Blockchain.
+
+The aim of this open source project is to reach beyond automated testing, into formal verification territory. This means using first-order logic to model and prove properties of the protocol. We are developing a formal model of the SBFT protocol using the Dafny language and framework, to formally prove interesting correctness properties of the SBFT consensus algorithm.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/README.md:8]: -> "The aim of this formal model is to document the protocol, and obtain stronger guarantees than simple testing."

File: docs/sbft-formal-model/README.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+## Run the image
+
+Run the image connected to your filesystem so you can edit in your OS, and then run Dafny from inside the docker container:
+
+```bash
+mkdir work
+cd work
+docker container run -t -i --mount src="`pwd`",target=/home/dafnyserver/work,type=bind --workdir /home/dafnyserver/work jonhdotnet/summer_school:1.1 /bin/bash
+git clone https://github.com/HristoStaykov/sbft-formal-verification-model.git
```
Comment (teoparvanov) [inline docs/sbft-formal-model/README.md:42]: let's reference the vmware/concord-bft repo here.

File: docs/sbft-formal-model/README.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+```bash
+Dafny program verifier finished with 10 verified, 0 errors
+```
+
+# Acknowledgements and references
+
+This project builds on top of the concepts and framework described in
+https://github.com/GLaDOS-Michigan/summer-school-2021<br>
+
+Special thanks to Jon Howell for his help and support along the way.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/README.md:75]: Link with Jon's GitHub alias.

File: docs/sbft-formal-model/application.s.dfy [hunk 16]
```diff
@@ -0,0 +1,63 @@
+//#title Host protocol
+//#desc Define the host state machine here: message type, state machine for executing one
+//#desc host's part of the protocol.
+
+// See exercise01.dfy for an English design of the protocol.
```
Comment (teoparvanov) [inline docs/sbft-formal-model/application.s.dfy:5]: Replace with Concord-BFT copyright snippet (2022).

**Review by teoparvanov — APPROVED** [review]
Looks great, thanks @HristoStaykov ! I just have some cosmetic/administrative comments below.

**Review by arc-vmware — APPROVED** [review]
Thanks @HristoStaykov , this PR is really great. 
LGTM!!! (G == Great)

**Pull request merged**